File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> chars;
        for(char c: t){
            if(chars.find(c) == chars.end()) chars[c] = 0;
            chars[c] += 1;
        }

        int left = 0;
        int right = 0;
        int min_len = INT_MAX;
        int start = 0;
        int count = 0;

        while( right < s.size()){
            if (chars.find(s[right]) != chars.end() ){
                chars[s[right]]--;
                if(chars[s[right]] >= 0) count++;
            }

            while(count == t.size()){
                min_len = min(right-left+1, min_len);
                start = min_len == right-left+1 ? left : start;
                if(chars.find(s[left]) != chars.end()){
                    chars[s[left]]++;
                    if(chars[s[left]] > 0) count--;
                }
                left++;
            }
            right++;
        }

        return min_len == INT_MAX ? "" : s.substr(start, min_len);
    }

};

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        # sliding window
        # check the window contain t or not, use hash map to store the 
        # frequency 
        # the compare time is o(1) becasue the t only contain upper and lower english letter
        # 
        # if yes shrink the left window until the window does not contain t
        # use two var result_substring_start and result_substring_len to record the anser
        # update it during shringkinog
        # time complexity o(n)

        result_str_start = 0
        result_str_len = float('inf')

        count_t = Counter(t)
        cur_count = defaultdict(int)

        left = 0

        def contain_t(cur_count):
            return all(cur_count.get(char, 0) >= count for char, count in count_t.items())


        for right in range(len(s)):
            cur_count[s[right]] += 1

            while contain_t(cur_count) and left <= right:
                window_str_len = right - left + 1

                if window_str_len < result_str_len:
                    result_str_len = window_str_len
                    result_str_start = left

                cur_count[s[left]] -= 1
                left += 1
        
        return s[result_str_start: result_str_start + result_str_len] if result_str_len != float('inf') else "" 




        
        